org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/hover/JvmAnnotationReferencePrinter.java

```diff
@@ -0,0 +1,220 @@
+
+		List<JvmAnnotationValue> explicitValues = reference.getExplicitValues();
+		boolean needsExplicitProperties = explicitValues.size() > 1 //
+				|| (!explicitValues.isEmpty() && explicitValues.get(0).getOperation() != null //
+						&& !"value".equals(explicitValues.get(0).getOperation().getSimpleName()));
+
+		if (!explicitValues.isEmpty()) {
+			buffer.append("(");
+			buffer.append(explicitValues.stream().map(explicitValue -> {
+				StringBuilder builder = new StringBuilder();
+				if (needsExplicitProperties) {
+					Iterable<JvmOperation> declaredOperations = reference.getAnnotation().getDeclaredOperations();
+					JvmOperation valueOperation = Iterables.tryFind(declaredOperations, o -> "value".equals(o.getSimpleName())).orNull();
+					JvmOperation operation = explicitValue.getOperation() != null ? explicitValue.getOperation() : valueOperation;
+					URI operationUri = EcoreUtil.getURI(operation);
+					builder.append(createLinkWithLabel(XtextElementLinks.XTEXTDOC_SCHEME, operationUri, operation.getSimpleName()));
+					builder.append("=");
+				}
+				builder.append(internalToString(explicitValue));
+				return builder.toString();
+			}).collect(Collectors.joining(", ")));
+			buffer.append(")");
+		}
+		return buffer.toString();
+	}
+
+	protected String _internalToString(XAnnotation reference) {
+		StringBuilder buffer = new StringBuilder();
+		buffer.append("@");
+		URI uri = EcoreUtil.getURI(reference.getAnnotationType());
+		buffer.append(createLinkWithLabel(XtextElementLinks.XTEXTDOC_SCHEME, uri, reference.getAnnotationType().getSimpleName()));
+
+		if (reference.getValue() != null) {
+			buffer.append("(");
+			buffer.append(reference.getValue());
+			buffer.append(")");
+		} else if (!reference.getElementValuePairs().isEmpty()) {
+			buffer.append("(");
+			buffer.append(reference.getElementValuePairs().stream().map(evPair -> {
+				StringBuilder builder = new StringBuilder();
+
+				Iterable<JvmOperation> declaredOperations = ((JvmAnnotationType) reference.getAnnotationType()).getDeclaredOperations();
+				JvmOperation valueOperation = Iterables.tryFind(declaredOperations, o -> "value".equals(o.getSimpleName())).orNull();
+				JvmOperation operation = evPair.getElement() != null ? evPair.getElement() : valueOperation;
+				URI operationUri = EcoreUtil.getURI(operation);
+				builder.append(createLinkWithLabel(XtextElementLinks.XTEXTDOC_SCHEME, operationUri, operation.getSimpleName()));
+				builder.append("=");
+				builder.append(internalToString(evPair.getValue()));
+				return builder.toString();
+			}).collect(Collectors.joining(", ")));
+			buffer.append(")");
+		}
+
+		return buffer.toString();
+	}
+
+	protected String _internalToString(JvmAnnotationValue jvmAnnotationValue) {
+		EStructuralFeature ref = jvmAnnotationValue.eClass().getEStructuralFeature("values");
+		if (ref == null) {
+			throw new IllegalStateException("Cannot find feature with name 'values' in " + jvmAnnotationValue);
```

Comment:
Updated the error message to provide more contextual information in case of failure.